### PR TITLE
[DEV-634] change user

### PR DESCRIPTION
**Pull request by Krysh90**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/user/models/user/user.service.ts [hunk 1]
```diff
@@ -64,6 +65,29 @@ export class UserService {
...
 
+  async getLinkedUser(id: number, address: string, blockchain: string): Promise<User> {
+    return this.userRepo
+      .createQueryBuilder('user')
+      .select('linkedUser.*')
+      .leftJoin('user.userData', 'userData')
+      .leftJoin('userData.users', 'linkedUser')
+      .where('user.id = :id', { id })
+      .andWhere('linkedUser.address = :address AND linkedUser.blockchain = :blockchain', { address, blockchain })
+      .getRawOne<User>();
```
Comment (davidleomay) [inline src/user/models/user/user.service.ts:76]: `.getOne()`?

File: src/user/models/user/user.service.ts [hunk 2]
```diff
@@ -64,6 +65,29 @@ export class UserService {
...
+  async getAllLinkedUsers(id: number): Promise<LinkedUserDto[]> {
+    return this.userRepo
+      .createQueryBuilder('user')
+      .select(
+        'linkedUser.address, linkedUser.blockchain, isSwitchable = CASE WHEN linkedUser.stakingBalance = 0 then 1 else 0 end',
+      )
+      .leftJoin('user.userData', 'userData')
+      .leftJoin('userData.users', 'linkedUser')
+      .where('user.id = :id', { id })
+      .getRawMany<LinkedUserDto>();
```
Comment (davidleomay) [inline src/user/models/user/user.service.ts:88]: We could replace this query by a simpler one and do the mapping to LinkedUser in Code. WDYT?

File: src/user/models/user/user.controller.ts [hunk 3]
```diff
@@ -42,6 +44,14 @@ export class UserController {
     return this.userService.updateUser(jwt.id, newUser);
   }
 
+  @Post('change')
+  @ApiBearerAuth()
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.USER))
+  async changeUser(@GetJwt() jwt: JwtPayload, @Body() changeUser: LinkedUserDto): Promise<{ accessToken: string }> {
+    const checkedUser = await this.userService.changeUser(jwt.id, changeUser);
+    return this.authService.signIn(checkedUser);
```
Comment (davidleomay) [inline src/user/models/user/user.controller.ts:52]: I would directly use the `AuthService.generateToken` method here.

File: src/user/models/user/dto/linked-user.dto.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsEnum(Blockchain)
+  blockchain: Blockchain;
+
+  @ApiPropertyOptional()
+  @IsBoolean()
+  isSwitchable: boolean;
+}
```
Comment (davidleomay) [inline src/user/models/user/dto/linked-user.dto.ts:21]: Would it make sense to split this in a in- and out-DTO? Might be disturbing to have the `isSwitchable` in the in-DTO.

File: src/user/models/user/user.service.ts (hunk 2 above)
Comment (Krysh90) [reply src/user/models/user/user.service.ts]: I am a little bit torn on this one. I would keep it like that and improve it in a later stage if needed?

File: src/user/models/user/user.service.ts (hunk 1 above)
Comment (Krysh90) [reply src/user/models/user/user.service.ts]: `.getOne()` doesn't return a user, whereby `.getRawOne<User>()` does.
Is this impacted by `SELECT linkedUser.*`?

File: src/user/models/user/user.controller.ts (hunk 3 above)
Comment (Krysh90) [reply src/user/models/user/user.controller.ts]: move `changeUser` to `AuthService`

File: src/user/models/user/dto/linked-user.dto.ts (hunk 4 above)
Comment (Krysh90) [reply src/user/models/user/dto/linked-user.dto.ts]: done

File: src/user/models/user/dto/linked-user.dto.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsEnum(Blockchain)
+  blockchain: Blockchain;
+}
+
+export class LinkedUserOutDto extends LinkedUserInDto {
+  @ApiPropertyOptional()
+  @IsBoolean()
```
Comment (davidleomay) [inline src/user/models/user/dto/linked-user.dto.ts:21]: `@ApiProperty()` (non nullable) and remove `@IsBoolean()` (no validation on out DTO)

**Review by davidleomay — APPROVED** [review]

File: src/user/models/user/dto/linked-user.dto.ts (hunk 5 above)
Comment (Krysh90) [reply src/user/models/user/dto/linked-user.dto.ts:21]: oh true - just copy pasted

**Pull request merged**